File: ai-orchestrator/src/main/java/com/lingxi/ai/orchestrator/service/TaskExecutionControl.java
package com.lingxi.ai.orchestrator.service;

import com.lingxi.ai.orchestrator.entity.Node;
import com.lingxi.ai.orchestrator.entity.NodeStatus;
import com.lingxi.ai.orchestrator.entity.Task;
import com.lingxi.ai.orchestrator.entity.TaskStatus;
import com.lingxi.ai.orchestrator.repository.NodeRepository;
import com.lingxi.ai.orchestrator.repository.TaskRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Slf4j
@Service
@RequiredArgsConstructor
public class TaskExecutionControl {

    private final TaskRepository taskRepository;
    private final NodeRepository nodeRepository;
    private final StateService stateService;
    private final RetryPolicy retryPolicy;

    public boolean canScheduleNodes(String taskId) {
        Optional<Task> taskOpt = taskRepository.findById(taskId);
        if (taskOpt.isEmpty()) {
            return false;
        }

        Task task = taskOpt.get();

        if (task.getStatus() == TaskStatus.PAUSED) {
            log.debug("Task {} is PAUSED, skipping scheduling", taskId);
            return false;
        }

        if (task.getStatus() == TaskStatus.SUCCESS || task.getStatus() == TaskStatus.FAILED) {
            return false;
        }

        List<Node> failedNodes = nodeRepository.findByTaskIdAndStatusIn(
                taskId,
                List.of(NodeStatus.FAILED)
        );

        if (!failedNodes.isEmpty()) {
            log.debug("Task {} has {} failed nodes, pausing scheduling",
                    taskId, failedNodes.size());
            return false;
        }

        return true;
    }

    @Transactional
    public void pauseTask(String taskId, String reason) {
        stateService.transitionTask(taskId, TaskStatus.PAUSED);
        log.info("Task {} paused, reason: {}", taskId, reason);
    }

    @Transactional
    public void resumeTask(String taskId) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new IllegalArgumentException("Task not found: " + taskId));

        if (task.getStatus() == TaskStatus.PAUSED) {
            stateService.transitionTask(taskId, TaskStatus.RUNNING);
            log.info("Task {} resumed", taskId);
        }
    }

    public boolean canRetryNode(String nodeId) {
        Node node = nodeRepository.findById(nodeId).orElse(null);
        if (node == null) {
            return false;
        }
        return node.getStatus() == NodeStatus.FAILED
                && retryPolicy.shouldRetry(node.getRetryCount(), node.getMaxRetry());
    }

    @Transactional
    public void retryNode(String nodeId) {
        Node node = nodeRepository.findById(nodeId)
                .orElseThrow(() -> new IllegalArgumentException("Node not found: " + nodeId));

        if (node.getStatus() != NodeStatus.FAILED) {
            throw new IllegalStateException("Node is not in FAILED status: " + nodeId);
        }

        if (!retryPolicy.shouldRetry(node.getRetryCount(), node.getMaxRetry())) {
            throw new IllegalStateException("Node has reached max retry count: " + nodeId);
        }

        node.setStatus(NodeStatus.RETRYING);
        node.setRetryCount(node.getRetryCount() + 1);
        node.setErrorMessage(null);
        nodeRepository.save(node);

        stateService.transitionNode(nodeId, NodeStatus.CREATED, null, null);

        log.info("Node {} marked for retry (attempt {}/{})",
                nodeId, node.getRetryCount(), node.getMaxRetry());

        if (node.getTaskId() != null) {
            Task task = taskRepository.findById(node.getTaskId()).orElse(null);
            if (task != null && task.getStatus() == TaskStatus.PAUSED) {
                resumeTask(node.getTaskId());
            }
        }
    }

    public String getTaskPauseReason(String taskId) {
        List<Node> failedNodes = nodeRepository.findByTaskIdAndStatus(taskId, NodeStatus.FAILED);
        if (!failedNodes.isEmpty()) {
            Node failedNode = failedNodes.get(0);
            return String.format("Node %s failed: %s (retry %d/%d)",
                    failedNode.getId(),
                    failedNode.getErrorMessage(),
                    failedNode.getRetryCount(),
                    failedNode.getMaxRetry());
        }
        return "Unknown reason";
    }
}

File: aios-core/internal/orchestrator/service/task_execution_control.go
package service

import (
	"context"
	"fmt"

	"go.uber.org/zap"

	"github.com/tangying-ai/aios-core/internal/model"
	"github.com/tangying-ai/aios-core/internal/model/repository"
)

type TaskExecutionControl struct {
	taskRepo     repository.TaskRepo
	nodeRepo     repository.NodeRepo
	stateService *StateService
}

func NewTaskExecutionControl(
	taskRepo repository.TaskRepo,
	nodeRepo repository.NodeRepo,
	stateService *StateService,
) *TaskExecutionControl {
	return &TaskExecutionControl{
		taskRepo:     taskRepo,
		nodeRepo:     nodeRepo,
		stateService: stateService,
	}
}

func (tc *TaskExecutionControl) PauseTask(ctx context.Context, taskID, reason string) error {
	task, err := tc.taskRepo.FindByID(ctx, taskID)
	if err != nil {
		return fmt.Errorf("failed to find task: %w", err)
	}
	if task == nil {
		return fmt.Errorf("task not found: %s", taskID)
	}

	task.Status = model.TaskPaused
	task.PauseReason = reason
	if err := tc.taskRepo.Save(ctx, task); err != nil {
		return fmt.Errorf("failed to pause task: %w", err)
	}
	zap.L().Info("Task paused", zap.String("taskId", taskID), zap.String("reason", reason))
	return nil
}

func (tc *TaskExecutionControl) ResumeTask(ctx context.Context, taskID string) error {
	task, err := tc.taskRepo.FindByID(ctx, taskID)
	if err != nil {
		return fmt.Errorf("failed to find task: %w", err)
	}
	if task == nil {
		return fmt.Errorf("task not found: %s", taskID)
	}

	task.Status = model.TaskRunning
	task.PauseReason = ""
	if err := tc.taskRepo.Save(ctx, task); err != nil {
		return fmt.Errorf("failed to resume task: %w", err)
	}
	zap.L().Info("Task resumed", zap.String("taskId", taskID))
	return nil
}

func (tc *TaskExecutionControl) FailTask(ctx context.Context, taskID string) error {
	task, err := tc.taskRepo.FindByID(ctx, taskID)
	if err != nil {
		return fmt.Errorf("failed to find task: %w", err)
	}
	if task == nil {
		return fmt.Errorf("task not found: %s", taskID)
	}

	return tc.stateService.TransitionTask(ctx, taskID, model.TaskFailed)
}

func (tc *TaskExecutionControl) RetryNode(ctx context.Context, nodeID string) error {
	node, err := tc.nodeRepo.FindByID(ctx, nodeID)
	if err != nil {
		return fmt.Errorf("failed to find node: %w", err)
	}
	if node == nil {
		return fmt.Errorf("node not found: %s", nodeID)
	}

	node.Status = model.NodeCreated
	node.RetryCount = 0
	node.ErrorMessage = ""
	if err := tc.nodeRepo.Save(ctx, node); err != nil {
		return fmt.Errorf("failed to save node: %w", err)
	}

	if err := tc.stateService.InitializeNodeReady(ctx, node); err != nil {
		return fmt.Errorf("failed to initialize node as ready: %w", err)
	}

	zap.L().Info("Node retry initiated", zap.String("nodeId", nodeID))
	return nil
}

func (tc *TaskExecutionControl) GetTaskPauseReason(ctx context.Context, taskID string) string {
	task, err := tc.taskRepo.FindByID(ctx, taskID)
	if err != nil || task == nil {
		return "Task not found"
	}
	if task.Status == model.TaskPaused {
		if task.PauseReason != "" {
			return task.PauseReason
		}
		return "Task was paused"
	}
	return "Task is not paused"
}
